test/run.py

```diff
@@ -0,0 +1,248 @@
+        pipe_pairs = [multiprocessing.Pipe() for _ in end_to_end_tests]
+        process_args = [
+            (
+                {
+                    'network': network,
+                    'plz_host': plz_host,
+                    'plz_port': plz_port,
+                    'bless': bless,
+                    'test_name': end_to_end_tests[i]
+                },
+                output_pipe
+            )
+            for i, (_, output_pipe) in enumerate(pipe_pairs)]
+        readers = [os.fdopen(r.fileno(), 'r') for r, _ in pipe_pairs]
+        pool_result = pool.starmap_async(
+            _run_test_piping_output,
+            process_args)
+
+        result_ready = False
+        while not result_ready:
+            if pool_result.ready():
+                result_ready = True
+            for reader in readers:
+                while True:
+                    line = reader.readline()
+                    if len(line) == 0:
+                        break
+                    print(line)
+        if not pool_result.successful():
+            test_utils.print_error('Some tests failed to run')
+            test_utils.print_error('Tests:', end_to_end_tests)
+            test_utils.print_error('Result:', pool_result.get())
+            return set(end_to_end_tests)
+
+        return set(t for i, t in enumerate(end_to_end_tests)
+                   if not pool_result.get()[i])
+    else:
+        failed_tests = set()
+        for test_name in end_to_end_tests:
+            success = run_end_to_end_test(
+                network, plz_host, plz_port, test_name, bless)
+            if not success:
+                failed_tests.add(test_name)
+        return failed_tests
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        'test_dirs', nargs='*', type=str, default=None)
+    parser.add_argument('--controller-tests-parameters', type=str, nargs='+',
+                        default=[])
+    parser.add_argument('--bless', action='store_true', default=False)
+    parser.add_argument('--in-parallel', action='store_true', default=False)
+    parser.add_argument('--end-to-end-only', action='store_true', default=False)
+    parser.add_argument('--controller-tests-only', action='store_true',
+                        default=False)
+    parser.add_argument('--clean-up-first', action='store_true', default=False)
+
+    options = parser.parse_args(sys.argv[1:])
+
+    if options.end_to_end_only and options.controller_tests_only:
+        raise ValueError('Options end_to_end_only and controller_tests_only '
+                         'can\'t be specified simultaneously')
+    end_to_end_only = options.end_to_end_only \
+        or len(options.test_dirs or []) > 0
+
+    end_to_end_tests = end_to_end_only or not options.controller_tests_only
+    controller_tests = options.controller_tests_only or not end_to_end_only
+
+    if options.clean_up_first:
+        test_utils.cleanup(interrupted=False)
```

Comment:
`interrupted` could have a default value of `False`.

Reply:
I'm not a fan of default values. They can be justified if you have a function with 53215 parameters and you're setting them most of the time to the same value and there are a lot of calls. Doesn't seem to be the case here.